src/java.base/share/native/libjava/io_util.c

```diff
@@ -57,6 +57,10 @@ readSingle(JNIEnv *env, jobject this, jfieldID fid) {
  */
 #define BUF_SIZE 8192
 
+/* The maximum size of a dynamically-allocated buffer.
+ */
+#define MAX_MALLOC_SIZE 2097152
```

Comment:
How did you choose 2Mb, it seems high.

Reply:
It's based on micro-benchmarks. Having the loops in Java reduces throughput but allocating memory using `malloc(len)` also reduces throughput as `len` gets larger and this threshold appears to balance the two.

Reply:
> It's based on micro-benchmarks. Having the loops in Java reduces throughput but allocating memory using `malloc(len)` also reduces throughput as `len` gets larger and this threshold appears to balance the two.

Are these micro benchmarks dropping the file system cache so there is real file I/O? I wasn't expecting to see a buffer larger than 1Mb so curious what the benchmarks say.

Reply:
There was no dropping of the file system cache. I would think this would have more of an effect on measuring the throughput of writing than reading.

I since modified the benchmark to invoke `FileChannel::force` in a `@TearDown` at the invocation level. That is not really copacetic but appears to give more believable results.

New runs show that if the malloc limit is 1Mb, then the revised code underperforms the current code until the size of array surpasses 1.5Mb. If the limit is 1.5Mb however, the new code improved throughput for all sizes tested. Based on this I reduced the malloc limit to 1572864.

These results are for macOS only thus far.

Reply:
> There was no dropping of the file system cache. I would think this would have more of an effect on measuring the throughput of writing than reading.

In that case, I assume the benchmarks are just reading from the file system cache so there isn't any actual file I/O.